### ISSUE TITLE
Variances can go (numerically) negative

### ISSUE THREAD

**Issue by lionfish0** [SheffieldML/GPy#253]
Mainly a problem for plotting.

Variances can become negative, which causes plots to 'flip'?

(unclear if the problem goes beyond plotting, ask Alan).

**Comment by mzwiessele** [comment]
Could you please provide an example script? So we can reproduce the error?

**Comment by mzwiessele** [comment]
I thought this was fixed actually, as there is always constant (infinitesimal) jitter on the inverses. Please provide some example, then we can check how that happens.

**Comment by alansaul** [comment]
Sorry I forgot about this until I ran into it again. This happens with numerical inaccuracies I believe, but in some cases its drastic and we should catch it (see example below).

Both noise variance and Gaussian noise are learnt to be very small, K will get numerically unstable (small negative eigenvalues in the following example). We should be able to deal with this when predicting, at least by adding jitter. In the example we below, we have (almost) identical observations, which will mean the matrix isn't full rank, or is very close to not being (minimum eigenvalue of 9.99999892865e-08).
To be honest, I'm not sure how this matrix is even being cholesky decomposed, as the jitter we add is not random, it is connected to the mean of the diagonal of K:
https://github.com/SheffieldML/GPy/blob/devel/GPy/util/linalg.py#L66
I have had similar effects occur when we don't have duplicate observations, I will try and find another example, so we need to figure out a way to resolve it for any case that K may become unstable.

Of course we could add a larger white variance, but I think we should at the very least catch these issues, and a best try and figure out how we can resolve them.

```
import matplotlib
import numpy as np
import GPy
import matplotlib.pyplot as plt
np.random.seed(0)
X = np.linspace(0,5)[:, None]
X = np.vstack((X, X[-10:,:]))
Y = np.sin(X) + np.random.randn(*X.shape)*1e-5

SE = GPy.kern.RBF(input_dim=1) + GPy.kern.White(1, variance=1e-7)
kernel = SE
m = GPy.models.GPRegression(X, Y, kernel)
m.kern.white.fix()
m.optimize()
m.plot()
mean, var = m._raw_predict(m.X)
print(var.min())
print(np.diag(m.kern.K(m.X)))
print(np.linalg.eigvals(m.kern.K(m.X)).min())
```

**Comment by mzwiessele** [comment]
We are adding a constant noise of 1e-8 which should be plenty for the Cholesky decomposition. We can of course check the variances after prediction and add noise to posterior.covariance and re apply the inverse. We can also solve the covariance directly into Kfu, so that the inverse is more stable?

**Comment by simbartonels** [comment]
During the GPSS summer school I noticed that the variances can be quite drastically negative. This is more than just numerical instabilities: -2.17720158e+08
Below is one of the examples a bit adapted.

```python
import numpy as np
import GPy

# set seed for reproducability
np.random.seed(3)

# Definition of the Branin test function
def branin(X):
    y = (X[:,1]-5.1/(4*np.pi**2)*X[:,0]**2+5*X[:,0]/np.pi-6)**2
    y += 10*(1-1/(8*np.pi))*np.cos(X[:,0])+10
    return(y)

# Training set defined as a 5*5 grid:
xg1 = np.linspace(-5,10,5)
xg2 = np.linspace(0,15,5)
X = np.zeros((xg1.size * xg2.size,2))
for i,x1 in enumerate(xg1):
    for j,x2 in enumerate(xg2):
        X[i+xg1.size*j,:] = [x1,x2]

Y = branin(X)[:,None]

# Fit a GP
# Create an exponentiated quadratic plus bias covariance function
k = GPy.kern.RBF(input_dim=2, ARD = True)

# Build a GP model
m = GPy.models.GPRegression(X,Y,k)

# fix the noise variance
m.likelihood.variance.fix(1e-5)

# Randomize the model and optimize
m.randomize()
m.optimize()

# Compute the mean of model prediction on 1e5 Monte Carlo samples
Xp = np.random.uniform(size=(1e5,2))
Xp[:,0] = Xp[:,0]*15-5
Xp[:,1] = Xp[:,1]*15
mu, var = m.predict(Xp)
print "The smallest encountered variance is: "
print min(var)
```

Best, Simon.

**Comment by mzwiessele** [comment]
Ah thank you very much!

The Branin function is designed to have changing gradients, depending on the input location.

Thus, the RBF function is not a good choice of kernel to describe it, as it assumes a uniform gradient, not depending on input location (infinitely smooth). 

Could you try using a Matern32 kernel? Also an Exponential kernel could be suited to describe, but not interpolate very nicely the problem at hand. When you fix the likelihood variance of a GP unexpected things can happen. This is comparable to a non-constrained ridge regression, which will do unexpected fittings. Thus, it is better to actually choose the right model of latent function, then forcing an RBF kernel to describe a not infinitely smooth function. 

Secondly, if there is no variance on the likelihood, numerical instabilities show drastically, therefore, -2e8 is not different from -2.

I hope this helps you understand the problem at hand. It is not the prediction (maths) doing something wrong, it is the underlying function, which implies assumptions on the training data.

**Comment by mzwiessele** [comment]
As for the example Alan gave, the better suited latent function is a periodic function. Thus, I would suggest trying a periodic kernel, which will be better suited to capture the underlying function. 

Does that help? 

So sometimes, it is easy to question the maths behind an ill posed problem. I am certain, that both these problems can be solved using the right latent function to explain both datasets. 

Please let me know if those suggestions help, so I can close the issue.

**Comment by lawrennd** [comment]
It's maybe a little misleading to worry about an assumption of 'uniform
gradient'. The RBF definitely assumes the function is stationary and
infinitely smooth, but the gradients vary across space.

We should get to the bottom of this numerical issue. @mosb has spoken about
possible solutions to these problems before, at the first GPSS, not sure if
they apply here.

See slide 49 of this talk:

http://gpss.cc/gpss13/assets/Sheffield-GPSS2013-Osborne.pdf

It may well be that Matern32 or Matern52 works better, but I just wanted to
rephrase the reasoning a bit in case it misled!

Neil

**Comment by mzwiessele** [comment]
We are already adding taking care of this. This issue seems to be related to the dataset and not to numerical instability. So, I tried to come up with a solution to the dataset, as an RBF seems to not be fitted for the function at hand. If I have misled the reader here I am very sorry. I hope that makes it more clear now. 

Also adding a bit of noise to the observations of the functions would shake the assumptions about the Branin function a little bit more loose. This could help as well.

**Comment by simbartonels** [comment]
Maybe my intuition about GPy is wrong. With the following call
```
m.likelihood.variance.fix(1e-5)
```
I just fix the noise to 1e-5, i.e. adding 1e-5 to the diagonal of my 25x25 kernel matrix, right?
If that's the case, then regardless of whether the model is ill-suited or not, shouldn't the predictive variance be positive? The parameters are really bad but the dataset is too small to blame this highly negative variance on numerical instabilities.

**Comment by zhenwendai** [comment]
According to my little investigation of this problem (BTW, thanks a lot for providing the example. It is very helpful.), it is due to our particular way of handling the inversion of covariance matrix. We calculate the inversion of the covariance matrix of data during the optimization, and store the value of this inversion for prediction. Although it is handy and computationally efficient, it is horrible in terms of numerical accuracy. We all know that solving an equation is much better than matrix inversion. For this particular example, we can solve this problem by handling the matrix inversion by solving on the fly. The following piece of code demonstrate the difference of these two approaches. Regarding GPy implementation, we probably have to sit down and discuss a more numerical stable and efficient approach.

```python
from scipy.linalg import solve
cov = m.kern.K(m.X)+np.eye(m.X.shape[0])*m.likelihood.variance.values
kfs = m.kern.K(m.X, Xp)
kss = m.kern.Kdiag(Xp)
var = kss -  (kfs*m.posterior.woodbury_inv.dot(kfs)).sum(0)
print var.min()
[out]: -726561.512757

from scipy.linalg import solve
cov = m.kern.K(m.X)+np.eye(m.X.shape[0])*m.likelihood.variance.values
kfs = m.kern.K(m.X, Xp)
kss = m.kern.Kdiag(Xp)
var = kss -  (kfs*solve(cov, kfs)).sum(0)
print var.min()
[out]: 5.54323196411e-06
```

**Comment by mzwiessele** [comment]
Very good point!!! Thank you so much for getting down to the problem at hand. 
With this in mind I thought about the structure of the posterior.
We also store the woodbury_chol in the posterior, So we can do:

from GPy.util.linalg import dtrtrs
cov = m.kern.K(m.X)+np.eye(m.X.shape[0])*m.likelihood.variance.values
kfs = m.kern.K(m.X, Xp)
kss = m.kern.Kdiag(Xp)
var = kss -  (kfs*GPy.util.linalg.dpotrs(m.posterior.woodbury_chol, kfs)[0]).sum(0)
print var.min() #Out[107]: 5.6326389312744141e-06
This should fix all the problems?

**Comment by mzwiessele** [comment]
```python
from GPy.util.linalg import dtrtrs
cov = m.kern.K(m.X)+np.eye(m.X.shape[0])*m.likelihood.variance.values
kfs = m.kern.K(m.X, Xp)
kss = m.kern.Kdiag(Xp)
var = kss -  (kfs*GPy.util.linalg.dpotrs(m.posterior.woodbury_chol, kfs)[0]).sum(0)
print var.min()
#Out[107]: 5.6326389312744141e-06
```

**Comment by mzwiessele** [comment]
fixed #314
